modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java

```diff
@@ -392,4 +461,526 @@ private void handleEncoderOutput(List<File> output, String message) {
+          int transitionDuration, boolean hasVideo, boolean hasAudio)
+          throws EncoderException, IllegalArgumentException {
+    logger.trace("MultiTrimConcat called with {} ", inputs, edits);
+    if (inputs == null || inputs.size() < 1) {
+      throw new IllegalArgumentException("At least one track must be specified.");
+    }
+    if (edits == null && inputs.size() > 1) {
+      throw new IllegalArgumentException("If there is no editing, only one track can be specified.");
+    }
+    List<VideoClip> clips = null;
+    if (edits != null) {
+      clips = new ArrayList<VideoClip>(edits.size() / 3);
+      int adjust = 0;
+      // When the first clip starts at 0, and there is a fade, lip sync can be off,
+      // this adjustment will mitigate the problem
+      for (int i = 0; i < edits.size(); i += 3) {
+        if (edits.get(i + 1) < transitionDuration) // If taken from the beginning of video
+          adjust = transitionDuration / 2000; // add half the fade duration in seconds
+        else
+          adjust = 0;
+        clips.add(new VideoClip(edits.get(i).intValue(), (double) edits.get(i + 1) / 1000 + adjust,
+              (double) edits.get(i + 2) / 1000));
+      }
+      try {
+        clips = sortSegments(clips, transitionDuration / 1000); // remove bad edit points
+      } catch (Exception e) {
+        logger.error("Illegal edits, cannot sort segment", e);
+      throw new EncoderException("Cannot understand the edit points", e);
+      }
+    }
+    // Set encoding parameters
+    Map<String, String> params = null;
+    if (inputs.size() > 0) { // Shared parameters - the rest are profile specific
+      params = getParamsFromFile(inputs.get(0));
+    }
+    if (profiles == null || profiles.size() == 0) {
+      logger.error("Missing encoding profiles");
+      throw new EncoderException("Missing encoding profile(s)");
+    }
+    try {
+      List<String> command = new ArrayList<>();
+      List<String> clauses = makeEdits(clips, transitionDuration, hasVideo, hasAudio); // map inputs into [ov]
+                                                                                               // and [oa]
+      // Entry point for multiencode here, if edits is empty, then use raw channels instead of output from edits
+      String videoOut = (clips == null) ? "[0:v]" : "[ov]";
+      String audioOut = (clips == null) ? "[0:a]" : "[oa]";
+      OutputAggregate outmaps = new OutputAggregate(profiles, params, (hasVideo ? videoOut : null),
+              (hasAudio ? audioOut : null)); // map outputs from ov and oa
+      if (hasAudio) {
+        clauses.add(outmaps.getAsplit());
+        clauses.add(outmaps.getAudioFilter());
+      }
+      if (hasVideo) {
+        clauses.add(outmaps.getVsplit());
+        clauses.add(outmaps.getVideoFilter());
+      }
+      while (clauses.remove(null)) {
+      } // remove all empty filters
```

Comment:
Not very important but maybe this is a bit cleaner:
```Java
clauses.removeIf(Objects::isNull);
```

Reply:
Updated